### PR TITLE
Avoid loading zip data into memory

### PR DESCRIPTION
**Pull request by shanemcd**
Link: https://github.com/ansible/ansible-runner/issues/604
Replaces: https://github.com/ansible/ansible-runner/pull/605

### CODE REVIEW

File: ansible_runner/utils/__init__.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 import uuid
 import codecs
 import zipfile
+import math
```
Comment (jbradberry) [inline ansible_runner/utils/__init__.py]: ???

File: ansible_runner/utils/streaming.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                    encoded_target.write(line)
+
+
+def unstream_dir(stream, length, target_directory):            
+    # NOTE: caller needs to process exceptions
+    with tempfile.NamedTemporaryFile() as tmp:
+        with open(tmp.name, "wb") as target:
+            with Base64IO(stream) as source:                
+                while True:
+                    remaining = length
```
Comment (jbradberry) [inline ansible_runner/utils/streaming.py]: this is in the loop, so it's basically always going to be getting reset to the full size

File: ansible_runner/utils/streaming.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                    chunk_size = 1024 * 1000  # 1 MB
+                    if chunk_size >= remaining:
+                        chunk_size = remaining
+                    data = source.read(chunk_size)
+                    target.write(data)
+                    remaining -= chunk_size
+
+                    if remaining == 0:
```
Comment (jbradberry) [inline ansible_runner/utils/streaming.py]: the end condition can be folded into the `while` statement

File: ansible_runner/utils/streaming.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            if source_directory:
+                for dirpath, dirs, files in os.walk(source_directory):
+                    relpath = os.path.relpath(dirpath, source_directory)
+                    if relpath == ".":
+                        relpath = ""
+                    for fname in files:
+                        archive.write(os.path.join(dirpath, fname), arcname=os.path.join(relpath, fname))
+            archive.close()
+
+        zip_size = Path(tmp.name).stat().st_size
```
Comment (jbradberry) [inline ansible_runner/utils/streaming.py:23]: wait, did you lose the calculation of tarball -> base64 size?

File: ansible_runner/utils/streaming.py (hunk 2 above)
Comment (shanemcd) [reply ansible_runner/utils/streaming.py]: Nice catch!

File: ansible_runner/utils/streaming.py (hunk 4 above)
Comment (shanemcd) [reply ansible_runner/utils/streaming.py:23]: Yep! It turns out that `base64io` is smart enough that when you say you need to read `n` number of bytes, it gives you that many bytes of _unencoded_ data. I thought that was pretty neat.

File: ansible_runner/utils/__init__.py (hunk 1 above)
Comment (shanemcd) [reply ansible_runner/utils/__init__.py]: Ooops.  Left over from a failed experiment.

**Comment by shanemcd** [comment]
recheck

**Comment by shanemcd** [comment]
recheck

**Review by jbradberry — APPROVED** [review]

**Pull request merged**